spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeRowsExec.scala

```diff
@@ -46,6 +48,16 @@ case class MergeRowsExec(
     output: Seq[Attribute],
     child: SparkPlan) extends UnaryExecNode {
 
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = {
+    // request a local sort by the row ID attrs if shuffle hash joins are enabled
+    // this is needed to co-locate matches for the same target row after the shuffle
+    if (performCardinalityCheck && !conf.preferSortMergeJoin) {
```

Comment:
It is only the shuffle hash join that needs to be fixed. In case of sort-merge joins, the incoming and outcoming records are already sorted by the attributes in the ON condition, which is sufficient.

Reply:
Joins implementations that involve a broadcast are handled differently too.

Reply:
Hello Anton, 

I have a small doubt, as per my understanding, at present not preferingSMJ doesn't implies SHJ being selected it's conjuncted with canBuildLocalHashMapBySize threshold as well as muchSmaller check's ([CP](https://github.com/apache/spark/blob/v3.2.0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala#L277)).

Also there are conf's such as "spark.sql.adaptive.maxShuffledHashJoinLocalMapThreshold" in AQE which even works slightly orthogonal by adding internal "PREFER_SHUFFLE_HASH" hint

Any pointer's how are we thinking of handling it or is the handling even required (appologies I am very new to iceberg)  ?

Reply:
Yeah, the problem is that only Spark knows which join type it is going to select and we have almost no control over it. I knew we won’t have a shuffle hash join if that config is set in the non-AQE path. I need to take a closer look at AQE. If it can add a shuffle hash join, then we probably have to always request a sort.

Reply:
Good point, @singhpk234.

Reply:
I checked AQE and I think it would be safer for us to always request a local sort. In many cases, the incoming records would be implicitly sorted so it shouldn't be too expensive.

Reply:
Thank you @aokolnychyi for checking, it makes perfect sense :) !!!